File: vanilla-task-advanced.js
class Tool {
    constructor(name, operation) {
      this.name = name;
      this.operation = operation;
      this.usageCount = 0;
    }
  
    run(input) {
      return this.operation(input);
    }
  }
  
  class Task {
    constructor(taskId, description, runFunction, tools = []) {
      this.taskId = taskId;
      this.description = description;
      this.runFunction = runFunction;
      this.tools = tools;
      this.toolLimits = {};
    }
  
    useTool(toolName, input) {
      const tool = this.tools.find((tool) => tool.name === toolName);
      if (!tool) {
        throw new Error(`No tool found with name: ${toolName}`);
      }
      if (this.toolLimits[toolName] && this.toolLimits[toolName] <= tool.usageCount) {
        throw new Error(`Usage limit exceeded for tool: ${tool.name} in task: ${this.description}`);
      }
      tool.usageCount++;
      return tool.run(input);
    }
  
    async execute() {
      console.log(`Starting task: ${this.description}`);
      let result = await this.runFunction();
      for (const tool of this.tools) {
        result = this.useTool(tool.name, result);
      }
      console.log(`Finished task: ${this.description}`);
      return result;
    }
  
    setToolLimit(toolName, limit) {
      this.toolLimits[toolName] = limit;
    }
  }
  
  class CustomProcess {
    constructor(tasks = [], isParallel = false) {
      this.tasks = tasks;
      this.isParallel = isParallel;
      this.executionHistory = [];
    }
  
    async run() {
      const results = [];
      console.log(`Running tasks ${this.isParallel ? 'in parallel' : 'sequentially'}...`);
      if (this.isParallel) {
        const promises = this.tasks.map(task => this.executeTask(task));
        results.push(...await Promise.all(promises));
      } else {
        for (const task of this.tasks) {
          const result = await this.executeTask(task);
          results.push(result);
        }
      }
      return results;
    }
  
    async executeTask(task) {
      try {
        const result = await task.execute();
        this.executionHistory.push(`Task executed: ${task.description}`);
        return result;
      } catch (error) {
        console.error(`Error executing task: ${task.description}`, error);
        return error;
      }
    }
  
    addTask(task, repetitions = 1) {
      for (let i = 0; i < repetitions; i++) {
        this.tasks.push(task);
      }
    }
  
    clearTasks() {
      this.tasks = [];
      this.executionHistory = [];
    }
  
    getExecutionHistory() {
      return [...this.executionHistory];
    }
  }
  
  class Agent {
    constructor() {
      this.failures = [];
    }
  
    async executeProcess(process) {
      const results = await process.run();
      this.failures.push(...results.filter(result => result instanceof Error));
      return results;
    }
  
    getFailures() {
      return [...this.failures];
    }
  }
  
  async function main() {
    const tool1 = new Tool("UPPER", (text) => text.toUpperCase());
    const task1 = new Task("id_1", "hello", () => new Promise(resolve => setTimeout(() => resolve("hello (async)"), 2000)), [tool1]);
    task1.setToolLimit(tool1.name, 3);
  
    const task2 = new Task("id_2", "world", () => new Promise(resolve => setTimeout(() => resolve("world (async)"), 2000)));
  
    console.log("Running tasks in parallel:");
    const myProcess = new CustomProcess([task1, task2], true);
    const agent = new Agent();
    const results = await agent.executeProcess(myProcess);
    console.log("Results:", results.filter(result => !(result instanceof Error)));
    console.log("Execution history:", myProcess.getExecutionHistory().join(" "));
    console.log("Failures:", agent.getFailures().join("\n"));
  
    console.log("\nRunning tasks sequentially:");
    myProcess.clearTasks();
    myProcess.addTask(task1, 3);
    myProcess.addTask(task2);
    myProcess.isParallel = false;
    const results2 = await agent.executeProcess(myProcess);
    console.log("Results:", results2.filter(result => !(result instanceof Error)));
    console.log("Execution history:", myProcess.getExecutionHistory().join(" "));
    console.log("Failures:", agent.getFailures().join("\n"));
  }
  
  main();

File: vanilla-task-advanced.py
import asyncio
from datetime import datetime

class Tool:
    def __init__(self, name, operation):
        self.name = name
        self.operation = operation
        self.usage_count = 0

    def run(self, input):
        return self.operation(input)

class Task:
    def __init__(self, task_id, description, run_function, tools=None):
        self.task_id = task_id
        self.description = description
        self.run_function = run_function
        self.tools = tools if tools else []
        self.tool_limits = {}

    async def use_tool(self, tool_name, input):
        tool = next((tool for tool in self.tools if tool.name == tool_name), None)
        if not tool:
            raise Exception(f"No tool found with name: {tool_name}")
        if tool_name in self.tool_limits and self.tool_limits[tool_name] <= tool.usage_count:
            raise Exception(f"Usage limit exceeded for tool: {tool.name} in task: {self.description}")
        tool.usage_count += 1
        return tool.run(input)

    async def execute(self):
        print(f"{datetime.now()} - Starting task: {self.description}")
        result = await self.run_function()
        for tool in self.tools:
            result = await self.use_tool(tool.name, result)
        print(f"{datetime.now()} - Finished task: {self.description}")
        return result

    def set_tool_limit(self, tool_name, limit):
        self.tool_limits[tool_name] = limit

class CustomProcess:
    def __init__(self, tasks=None, is_parallel=False):
        self.tasks = tasks if tasks else []
        self.is_parallel = is_parallel
        self.execution_history = []

    async def run(self):
        results = []
        print(f"{datetime.now()} - Running tasks {'in parallel' if self.is_parallel else 'sequentially'}...")
        if self.is_parallel:
            tasks = [self.execute_task(task) for task in self.tasks]
            results = await asyncio.gather(*tasks, return_exceptions=True)
        else:
            for task in self.tasks:
                result = await self.execute_task(task)
                results.append(result)
        return results

    async def execute_task(self, task):
        try:
            result = await task.execute()
            self.execution_history.append(f"Task executed: {task.description}")
            return result
        except Exception as error:
            print(f"{datetime.now()} - Error executing task: {task.description}", error)
            return error

    def add_task(self, task, repetitions=1):
        self.tasks.extend([task]*repetitions)

    def clear_tasks(self):
        self.tasks.clear()
        self.execution_history.clear()

    def get_execution_history(self):
        return self.execution_history.copy()

class Agent:
    def __init__(self):
        self.failures = []

    async def execute_process(self, process):
        results = await process.run()
        self.failures.extend(str(result) for result in results if isinstance(result, Exception))
        return results

    def get_failures(self):
        return self.failures.copy()

async def main():
    tool1 = Tool("UPPER", lambda text: text.upper())
    task1 = Task("id_1", "hello", lambda: asyncio.sleep(2, "hello (async)"), [tool1])
    task1.set_tool_limit(tool1.name, 3)

    task2 = Task("id_2", "world", lambda: asyncio.sleep(2, "world (async)"))

    print("Running tasks in parallel:")
    my_process = CustomProcess([task1, task2], True)
    agent = Agent()
    results = await agent.execute_process(my_process)
    print("Results:", [result for result in results if not isinstance(result, Exception)])
    print("Execution history:", " ".join(my_process.get_execution_history()))
    print("Failures:", "\n".join(agent.get_failures()))

    print("\nRunning tasks sequentially:")
    my_process.clear_tasks()
    my_process.add_task(task1, 3)
    my_process.add_task(task2)
    my_process.is_parallel = False
    results2 = await agent.execute_process(my_process)
    print("Results:", [result for result in results2 if not isinstance(result, Exception)])
    print("Execution history:", " ".join(my_process.get_execution_history()))
    print("Failures:", "\n".join(agent.get_failures()))

asyncio.run(main())
